src/script/components/LexicalInput/utils/mention-utils.ts

```diff
@@ -0,0 +1,204 @@
+        matchingString: matchingString,
+        replaceableString: matchingStringWithTrigger,
+      };
+    }
+  }
+  return null;
+}
+
+export function isWordChar(char: string, triggers: string[]) {
+  return new RegExp(VALID_CHARS(triggers)).test(char);
+}
+
+type SelectionInfo = {
+  node: LexicalNode;
+  offset: number;
+  isTextNode: boolean;
+  textContent: string;
+  selection: RangeSelection;
+  prevNode: LexicalNode | null;
+  nextNode: LexicalNode | null;
+  cursorAtStartOfNode: boolean;
+  cursorAtEndOfNode: boolean;
+  wordCharBeforeCursor: boolean;
+  wordCharAfterCursor: boolean;
+};
+
+export function getSelectionInfo(triggers: string[]): SelectionInfo | undefined {
+  const selection = $getSelection();
+  if (!selection || !$isRangeSelection(selection)) {
+    return undefined;
+  }
+
+  const anchor = selection.anchor;
+  const focus = selection.focus;
+  const nodes = selection.getNodes();
+  if (anchor.key !== focus.key || anchor.offset !== focus.offset || nodes.length === 0) {
+    return undefined;
+  }
+
+  const [node] = nodes;
+  const isTextNode = $isTextNode(node) && node.isSimpleText();
+  const offset = anchor.type === 'text' ? anchor.offset : 0;
+  const textContent = node.getTextContent();
+  const cursorAtStartOfNode = offset === 0;
+  const cursorAtEndOfNode = textContent.length === offset;
+  const charBeforeCursor = textContent.charAt(offset - 1);
+  const charAfterCursor = textContent.charAt(offset);
+  const wordCharBeforeCursor = isWordChar(charBeforeCursor, triggers);
+  const wordCharAfterCursor = isWordChar(charAfterCursor, triggers);
+  const prevNode = getPreviousSibling(node);
+  const nextNode = getNextSibling(node);
+
+  return {
+    node,
+    offset,
+    isTextNode,
+    textContent,
+    selection,
+    prevNode,
+    nextNode,
+    cursorAtStartOfNode,
+    cursorAtEndOfNode,
+    wordCharBeforeCursor,
+    wordCharAfterCursor,
+  };
+}
+
+export function insertMention(triggers: string[], trigger: string, value?: string) {
+  const selectionInfo = getSelectionInfo(triggers);
+  if (!selectionInfo) {
+    return false;
+  }
+
+  const {
+    node,
+    offset,
+    selection,
+    wordCharBeforeCursor,
+    wordCharAfterCursor,
+    cursorAtStartOfNode,
+    cursorAtEndOfNode,
+    prevNode,
+    nextNode,
+  } = selectionInfo;
+
+  // Insert a mention node or a text node with the trigger to open the mention menu.
+  const mentionNode = value ? $createBeautifulMentionNode(trigger, value) : $createTextNode(trigger);
+
+  // Insert a mention with a leading space if the node at the cursor is not a text node.
+  if (!($isParagraphNode(node) && offset === 0) && !$isTextNode(node)) {
+    selection.insertNodes([$createTextNode(' '), mentionNode]);
+    return true;
+  }
+
+  let spaceNode: TextNode | null = null;
+  const nodes: LexicalNode[] = [];
+  if (wordCharBeforeCursor || (cursorAtStartOfNode && prevNode !== null && !$isTextNode(prevNode))) {
```

Comment:
do we need to check wordCharBeforeCursor? since we are already checking cursorAtStartOfNode.

Reply:
still the same
```

 if (wordCharBeforeCursor || (cursorAtStartOfNode && prevNode !== null && !$isTextNode(prevNode))) {
    nodes.push($createTextNode(' '));
  }
```